src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs

```diff
@@ -1208,82 +1122,126 @@ public static string GetPSModulePath(PSModulePathScope scope)
 
-            // If the variable isn't set, then set it to the default value
-            if (currentProcessModulePath == null)  // EVT.Process does Not exist - really corner case
+            List<string> finalModulePath = new List<string>();
+            AddPath(finalModulePath, personalModulePathToUse);
+            AddPath(finalModulePath, sharedModulePath);
+            AddPath(finalModulePath, systemModulePathToUse);
+
+            // If the existing module path if provided, we add it to the end of the list.
+            if (currentProcessModulePath != null)
             {
-                // Handle the default case...
-                if (string.IsNullOrEmpty(hkcuUserModulePath)) // EVT.User does Not exist -> set to <SpecialFolder.MyDocuments> location
-                {
-                    currentProcessModulePath = personalModulePath; // = SpecialFolder.MyDocuments + Utils.ProductNameForDirectory + Utils.ModuleDirectory
-                }
-                else // EVT.User exists -> set to EVT.User
-                {
-                    currentProcessModulePath = hkcuUserModulePath; // = EVT.User
-                }
+                AddPath(finalModulePath, currentProcessModulePath);
+            }
+
+            return string.Join(Path.PathSeparator, finalModulePath);
+        }
 
-                if (string.IsNullOrEmpty(currentProcessModulePath))
+        /// <summary>
+        /// Adds the path if not present to the final module path list. Each
+        /// path is split by the path separator and added to the final list.
+        /// </summary>
+        /// <param name="finalModulePath">The final module path list.</param>
+        /// <param name="path">The path to add.</param>
+        private static void AddPath(List<string> finalModulePath, ReadOnlySpan<char> path)
+        {
+            ReadOnlySpan<char> dirSeparatorChars = stackalloc char[] {
+                Path.DirectorySeparatorChar,
+                Path.AltDirectorySeparatorChar
+            };
+
+#if UNIX
+            IEqualityComparer<char>? valueComparer = null;
+#else
+            IEqualityComparer<char> valueComparer = new WindowsPathEnvComparer();
+#endif
+
+            foreach (Range segment in path.Split(Path.PathSeparator))
+            {
+                // We trim any surrounding whitespace and trailing directory separators
+                // to canonicalize the path. The comparer on Windows will also treat
+                // '/' and '\' as the same and in a case insensitive operation.
+                scoped ReadOnlySpan<char> entry = path[segment].Trim();
+                if (entry.Length == 0)
                 {
-                    currentProcessModulePath ??= string.Empty;
+                    continue;
                 }
+#if UNIX
+                // Don't trim if it's just '/' as that's a valid entry on *nix.
+                else if (!(entry.Length == 1 && entry[0] == Path.DirectorySeparatorChar))
```

Comment:
Can we use [IO.Path.[IsPathRooted](https://source.dot.net/System.Private.CoreLib/R/ac667d4ce60e4fec.html)](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/IO/Path.Unix.cs,132)?

And what about `C:\` on Windows?

Reply:
I've updated the code to handle Windows absolute paths and their trailing chars and expanded the comments that explains what it does. I didn't use `TrimEndingDirectorySeparator()` because it will only trim 1 char at the end.